src/relay/pass/quantize_rewrite.cc

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file quantize_rewrite.cc
+ * \brief Lower quantized ops to exisiting Relay ops.
+ */
+
+#include <tvm/relay/pass.h>
+#include <tvm/relay/op_attr_types.h>
+#include <tvm/relay/attrs/nn_quantize.h>
+#include "pattern_util.h"
+
+namespace tvm {
+namespace relay {
+
+Expr QuantizedConv2DForwardRewrite(const Call& ref_call,
+                                   const Array<Expr>& new_args,
+                                   const NodeRef& ctx) {
+  // TODO(janimesh) - This is not the right calculation. This only serves as a
+  // prototype to discuss the flow of lowering of quantization ops and
+  // namespaces.
+  CHECK_EQ(new_args.size(), 2);
+  Expr quantized_data = new_args[0];
+  Expr quantized_kernel = new_args[1];
+  const auto* param = ref_call->attrs.as<QuantizedConv2DAttrs>();
+  CHECK_EQ(param->input_zero_point, 0) << "Only symmetric support yet";
+  CHECK_EQ(param->kernel_zero_point, 0) << "Only symmetric support yet";
+  CHECK_EQ(param->output_zero_point, 0) << "Only symmetric support yet";
+  // TODO(janimesh) - The out_dtype should be something else, like "int32".
+  Expr int8_conv = Conv2D(quantized_data,
+                          quantized_kernel,
+                          param->strides,
+                          param->padding,
+                          param->dilation,
+                          param->groups,
+                          param->channels,
+                          param->kernel_size,
+                          param->data_layout,
+                          param->kernel_layout,
+                          param->out_layout,
+                          Int(32));
+  // TODO(janimesh) - The out_dtype should come from outside..
+  int8_conv = Cast(int8_conv, param->out_dtype);
```

Comment:
Maybe we should name it `q_conv2d_requant` or something is better. Here is not just simple cast, we will need input scale / kernel scale / output scale / output min / output max and so on to compute the correct uint8 value. Just use Cast will confuse us here is just cast simply.

Reply:
Yes, I agree. This computation is not complete yet (I added a comment before but I dont thing it is visible enough).

I am looking into converting the scale computations into integer computations. I am having difficulty in understanding the rounding computations. Can you help with that?

Reply:
You may take [TFLite](https://github.com/tensorflow/tensorflow/blob/v2.0.0-beta1/tensorflow/lite/kernels/internal/reference/integer_ops/conv.h#L112) (all stuff in that function) as reference.

Reply:
Thanks, I will have a closer look this week and update the patch with the correct computation.